File: benchmark.py
SIZE = 50000
EXPECTED_CHECKSUM = 41667916675000

# Step 1: Initialize Arrays
A = [i for i in range(SIZE)]
B = [SIZE - i for i in range(SIZE)]
# Step 2: Reuse C to store prefix sums of B
C = [0] * SIZE
C[0] = B[0]

for i in range(1, SIZE):
    C[i] = C[i - 1] + B[i]

# Step 3: Compute checksum
checksum = sum(C)

# Step 4: Output the checksum
print(f"Checksum: {checksum}")

# Step 5: Check if checksum matches expected value
if checksum != EXPECTED_CHECKSUM:
    print(f"Error: Checksum mismatch! Expected {EXPECTED_CHECKSUM} but got {checksum}.")
    exit(1)

exit(0)

File: benchmark.c
#include <stdio.h>

#define SIZE 50000
#define EXPECTED_CHECKSUM 41667916675000ULL

// Statical allocation of arrays
int A[SIZE];
int B[SIZE];
unsigned long long C[SIZE]; // Use unsigned long long to prevent overflow

int main() {
    // Step 1: Initialize arrays
    // A[i] = i
    // B[i] = 50000 - i
    // C[i] = 0
    for (int i = 0; i < SIZE; i++) {
        A[i] = i;
        B[i] = SIZE - i;
        C[i] = 0ULL;
    }
    C[0] = B[0];


    // Step 2: For each element in C, sum all elements in B up to and including i
    for (int i = 1; i < SIZE; i++) {
        C[i] = C[i-1] + (unsigned long long)B[i];
    }

    // Step 3: Compute checksum by summing all elements in C
    unsigned long long checksum = 0ULL;
    for (int i = 0; i < SIZE; i++) {
        checksum += C[i];
    }

    // Step 4: Output the checksum
    printf("Checksum: %llu\n", checksum);

    // Step 5: Compare the checksum to the expected value
    if (checksum != EXPECTED_CHECKSUM) {
        fprintf(stderr, "Error: Checksum mismatch! Expected %llu but got %llu.\n",
                EXPECTED_CHECKSUM, checksum);
        return 1; // Non-zero return value indicates failure
    }

    return 0; // Success
}